### PR TITLE
[Java] [Feature] Load from SavedModel

### LINKED ISSUE #7134: [Java] [Feature] Load from SavedModel
Make it possible for the Java library to load exported TF models in SavedModel format.   

Outline of the proposed design:
1. Define a new class `SavedModelBundle` to represent the loaded saved model.  Properties include the graph, session, and metagraphdef.
2. Define a static method on `SavedModelBundle` to load from disk, given an export path and some tags.

```
class SavedModelBundle implements AutoCloseable { 
  public Graph graph() { ... }
  public Session session() { ... }
  public byte[] metaGraphDef() { ... }
  public void close() { /* close the session and graph */ }
  public static SavedModelBundle loadSavedModel(String exportDir, Set<String> tags) { ... }
}
```

3. Write a JNI method to support loading, based on `TF_LoadSessionFromSavedModel` from the C API.

### PR DESCRIPTION
**Pull request by EronWright**
Added `SavedModelBundle` class to represent a loaded saved model.

Closes #7134

### CODE REVIEW

**Comment by EronWright** [comment]
@asimshankar as we discussed.

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Perhaps mention that the session has been initialized with any saved tensor values.
Something like:

Returns the Session with state (e.g., values of tensors) initialized from state saved in the directory provided to {@code load}.

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Let's add a little more descriptive detail about SavedModels. Perhaps something like (with a link to MetaGraphDef):

/**
 * SavedModel represents a model loaded from storage.
 *
 * The model consists of a description of the computation (a {@link Graph}) , a {@link Session}
 * with tensors (e.g., parameters or variables in the graph) initialized to values saved on storage
 * and a description (<a href="https://www.tensorflow.org/code/tensorflow/core/protobuf/meta_graph.proto">MetaGraphDef</a>) of the model.
 */

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Returns the graph that describes the computation performed by the model.

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Nit: Can combine this into one:

/** Releases resources (the Graph and Session) of the model. */

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Can this be private? Also, mention in the comment that this is invoked from the native 'load' method.

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc
Comment (asimshankar) [inline tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: Nit: 

} else {

**Review by asimshankar — COMMENTED** [review]
Thanks for the PR @EronWright ! Much appreciated.

@jhseu will also be interested.

Some minor changes, mostly around naming and comments, but otherwise looks good. Thanks!

**Comment by jhseu** [comment]
Thanks for writing this! :+1:

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (EronWright) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: I avoided using the class name `SavedModel` because the def within `core/protobuf/saved_model.proto` is named `SavedModel`.

Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Ah, righto. And even though that is in a separate package, it would mean that any code using both will have to spell out the fully qualified class name (`org.tensorflow.SavedModel` vs `org.tensorflow.framework.SavedModel`). 

Okay, let's keep this as `SavedModelBundle` then. Thanks.

**Comment by EronWright** [comment]
Updated based on review comments.

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc
Comment (EronWright) [inline tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: I copied this snippet from `graph_jni.cc` so I'd prefer you follow up on that, OK?

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: This can be private too?

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc
Comment (asimshankar) [inline tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: Can we do away with all this MakeUniqueBuffer fanciness and instead have:

```
TF_Buffer* crun_options = nullptr;
// ...
if (crun_options != nullptr) TF_DeleteBuffer(crun_options);
```

Since there is only one place where we create the TF_Buffer and only one path to the point where it can be deleted, just like all other cleanup functions?

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Remove this line? (The first line below is just as good?)

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Use {@code ... } instead of <i>...</i>.

Or better still, since this is a private method and thus won't appear in Javadocs can do away with any annotations, which will make it easier to read in code.

.e.g, "handle to the C TF_Graph object and ..."

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc
Comment (asimshankar) [inline tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: Sounds fair.

**Review by asimshankar — COMMENTED** [review]
Thanks. Mind indulging in 3 more minor things and then let's get this in. Once again, thanks for the contribution!

**Review by asimshankar — APPROVED** [review]
Thanks!

**Comment by jhseu** [comment]
Jenkins, test this please

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc
Comment (asimshankar) [inline tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: I just realized that this will fail on the Android tests since we've currently disabled `TF_LoadSessionFromSavedModel` in Android (since it pulls it a bit too many dependencies right now).
I'll make a change to `c_api.h` so that the symbol will exist, even if the call will fail on Android.

So no changes needed in this PR for that, but just a heads up.

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by caisq** [comment]
@EronWright there are test failures: 

ERROR: /workspace/tensorflow/workspace.bzl:422:3: in java_import_external rule //external:junit: 
Traceback (most recent call last):
	File "/workspace/tensorflow/workspace.bzl", line 422
		java_import_external(name = 'junit')
	File "/var/lib/jenkins/workspace/tensorflow-pull-requests-gpu/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/external/io_bazel_rules_closure/closure/private/java_import_external.bzl", line 29, in _java_import_external
		repository_ctx.name
Object of type 'ctx' has no field "name".
ERROR: Analysis of target '//tensorflow/java:SavedModelBundleTest' failed; build aborted.

**Comment by jhseu** [comment]
The error seems unrelated to this change, but possibly related to other recent changes.

@jart, mind looking at the java_import_external error?

File: tensorflow/java/BUILD
Comment (jhseu) [inline tensorflow/java/BUILD]: This changed recently and you'll have to update it to say `@junit`.

**Comment by jhseu** [comment]
Ah, nevermind @jart, ignore the above. It looks like this change needs updating.

File: tensorflow/java/BUILD
Comment (jhseu) [inline tensorflow/java/BUILD]: (Note that you'll have to merge master or rebase first).

**Comment by jhseu** [comment]
Tip for future pull requests: it's a little easier if the `Allow edits from maintainers` option is left enabled, because there are often minor edits that we have to make before merging and we can often do them ourselves :)

**Comment by EronWright** [comment]
@jhseu "allow edits" was enabled on this PR but I'm happy with any approach that gets the patch in. :)

Closing in favor of #8148

**Comment by jhseu** [comment]
Weird, I tried pushing to that branch and it got rejected. Maybe it's because it was in cookieai rather than your personal repo?

**Pull request closed without merging**
